freeipa/src/main/java/com/sequenceiq/freeipa/events/FreeIpaCustomCrnOrNameProvider.java

```diff
@@ -0,0 +1,54 @@
+package com.sequenceiq.freeipa.events;
+
+import static com.sequenceiq.cloudbreak.structuredevent.rest.urlparser.CDPRestUrlParser.RESOURCE_ID;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.auth.ThreadBasedUserCrnProvider;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPOperationDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.rest.RestCallDetails;
+import com.sequenceiq.cloudbreak.structuredevent.rest.CustomCrnOrNameProvider;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Component
+public class FreeIpaCustomCrnOrNameProvider implements CustomCrnOrNameProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaCustomCrnOrNameProvider.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Override
+    public void provide(RestCallDetails restCallDetails, CDPOperationDetails operationDetails, Map<String, String> restParams,
+            String nameField, String crnField) {
+        try {
+            String requestUri = restCallDetails.getRestRequest().getRequestUri();
+            URI uri = new URI(requestUri);
+            List<NameValuePair> queryParams = URLEncodedUtils.parse(uri, StandardCharsets.UTF_8);
+            Optional<NameValuePair> environmentValue = queryParams.stream().filter(p -> "environment".equals(p.getName())).findFirst();
```

Comment:
unfortunately only `FreeIpaV1Endpoint` has `environment` in the URI, LDAP and Kerberos related endpoints have `environmentCrn`. I think those should be handled alse

Reply:
I love the consistency